rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPool.java

```diff
@@ -0,0 +1,171 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Daniel
+ */
+public class ObjectPool<T>
+{
+  private final PoolConfig config;
+  private final ObjectFactory<T> factory;
+  private final ObjectPoolPartition<T>[] partitions;
+  private Scavenger scavenger;
+  private volatile boolean shuttingDown;
+
+  public ObjectPool(PoolConfig poolConfig, ObjectFactory<T> objectFactory)
+  {
+    this.config = poolConfig;
+    this.factory = objectFactory;
+    this.partitions = new ObjectPoolPartition[config.getPartitionSize()];
+    if (config.getScavengeIntervalMilliseconds() > 0) {
+      this.scavenger = new Scavenger();
+      this.scavenger.start();
+    }
+  }
+
+  public void registerHost(String host, int socketTimeout, int connectTimeout, int index)
+  {
+    partitions[index] = new ObjectPoolPartition<>(this, index, config, factory, createBlockingQueue(config), host, socketTimeout, connectTimeout);
+  }
+
+  protected BlockingQueue<Poolable<T>> createBlockingQueue(PoolConfig poolConfig)
+  {
+    return new ArrayBlockingQueue<>(poolConfig.getMaxSize());
+  }
+
+  public Poolable<T> borrowObject(int partitionNumber)
+          throws SocketException
+  {
+    return borrowObject(true, partitionNumber);
+  }
+
+  public Poolable<T> borrowObject(boolean blocking, int partitionNumber)
+          throws SocketException
+  {
+    Log.info("Borrowing object for partition: ", partitionNumber);
+    for (int i = 0; i < 3; i++) { // try at most three times
+      Poolable<T> result = getObject(blocking, partitionNumber);
+      if (factory.validate(result.getObject())) {
+        return result;
+      }
+      else {
+        this.partitions[result.getPartition()].decreaseObject(result);
+      }
+    }
+    throw new RuntimeException("Cannot find a valid object");
+  }
+
+  private Poolable<T> getObject(boolean blocking, int partitionNumber)
+  {
+    if (shuttingDown) {
+      throw new IllegalStateException("Your pool is shutting down");
+    }
+    int partition = partitionNumber % this.config.getPartitionSize();
+    ObjectPoolPartition<T> subPool = this.partitions[partition];
+
+    Poolable<T> freeObject;
+    if (subPool.getObjectQueue().size() == 0) {
+      // increase objects and return one, it will return null if reach max size
+      subPool.increaseObjects(this.config.getDelta());
+      try {
+        if (blocking) {
+          Log.warn("Pool is empty...Waiting for connection");
+          freeObject = subPool.getObjectQueue().take();
+        }
+        else {
+          freeObject = subPool.getObjectQueue().poll(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
+          if (freeObject == null) {
+            throw new RuntimeException("Cannot get a free object from the pool");
+          }
+        }
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException(e); // will never happen
+      }
+    }
+    else {
+      freeObject = subPool.getObjectQueue().poll();
```

Comment:
There is a race here due to which we can return 'null' and hit NPE up in the stack.
When `subPool.getObjectQueue().size()` was checked, size was non-zero so we came to else but on execution of `freeObject = subPool.getObjectQueue().poll();` we can get null if another thread in parallel did a poll.

We should refactor this method to this:
`
private Poolable<T> getObject(boolean blocking, int partition)
  {
    if (shuttingDown) {
      throw new IllegalStateException("Your pool is shutting down");
    }
    ObjectPoolPartition<T> subPool = this.partitions[partition];

    Poolable<T> freeObject;
    if (subPool.getObjectQueue().size() == 0) {
      // increase objects and return one, it will return null if reach max size
      subPool.increaseObjects(this.config.getDelta());
    }
    
    try {
      if (blocking) {
        Log.warn("Pool is empty...Waiting for connection");
        freeObject = subPool.getObjectQueue().take();
      }
      else {
        freeObject = subPool.getObjectQueue().poll(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
        if (freeObject == null) {
          throw new RuntimeException("Cannot get a free object from the pool");
        }
      }
    }
    catch (InterruptedException e) {
      throw new RuntimeException(e); // will never happen
    }
    
    freeObject.setLastAccessTs(System.currentTimeMillis());
    return freeObject;
  }
`